File: CSharp/SpaceVIL/SpaceVIL/Items/Grid.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;
using SpaceVIL.Common;
using SpaceVIL.Core;
using System.Threading;

namespace SpaceVIL
{
    public class Grid : Prototype, IFreeLayout
    {
        static int count = 0;

        /// <summary>
        /// Constructs a Grid
        /// </summary>
        private Grid()
        {
            SetItemName("Grid_" + count);
            count++;
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.Grid)));
            IsFocusable = false;
        }

        /// <summary>
        /// Constructs a Grid with the given rows and columns
        /// </summary>
        public Grid(int rows, int columns) : this()
        {
            _rowCount = rows;
            _columnCount = columns;
            InitCells();
        }

        //rows and counts
        private List<Cell> _cells;
        internal void InitCells()
        {
            _cells = new List<Cell>();
            for (int i = 0; i < _rowCount; i++)
            {
                for (int j = 0; j < _columnCount; j++)
                {
                    _cells.Add(new Cell(this, i, j));
                }
            }
        }
        public void SetFormat(int rows, int columns)
        {
            if (rows == _rowCount && columns == _columnCount)
                return;

            _rowCount = rows;
            _columnCount = columns;
            RearrangeCells();
        }
        private void RearrangeCells()
        {
            if (_cells == null)
            {
                InitCells();
                return;
            }

            List<IBaseItem> items = new List<IBaseItem>();
            foreach (var cell in _cells)
                items.Add(cell.GetItem());
            InitCells();
            int index = 0;
            foreach (var item in items)
            {
                _cells.ElementAt(index).SetItem(item);
                index++;
                if (_cells.Count == index)
                    break;
            }
            UpdateLayout();
        }
        private int _rowCount = 1;

        /// <summary>
        /// Set count of the rows
        /// </summary>
        public void SetRowCount(int capacity)
        {
            if (capacity != _rowCount)
                _rowCount = capacity;
            RearrangeCells();
        }
        public int GetRowCount()
        {
            return _rowCount;
        }
        private int _columnCount = 1;

        /// <summary>
        /// Set count of the columns
        /// </summary>
        public void SetColumnCount(int capacity)
        {
            if (capacity != _columnCount)
                _columnCount = capacity;
            //Need to InitCells REFACTOR!
            RearrangeCells();
        }
        public int GetColumnCount()
        {
            return _columnCount;
        }

        /// <summary>
        /// Returns the cell by row and column number
        /// </summary>
        public Cell GetCell(int row, int column)
        {
            Cell cell = null;
            try
            {
                cell = _cells[column + row * _columnCount];
            }
            catch (Exception ex)
            {
                Console.WriteLine("Cells row and colums out of range.\n" + ex.ToString());
                return cell;
            }
            return cell;
        }

        /// <returns> all cells list </returns>
        public List<Cell> GetAllCells()
        {
            return _cells;
        }

        public override bool RemoveItem(IBaseItem item)
        {
            bool baseBool = base.RemoveItem(item);
            foreach (var link in _cells)
            {
                if (link.GetItem() == item)
                    link.SetItem(null);
            }
            // UpdateLayout();
            return baseBool;
        }

        public void RemoveItem(int row, int column)
        {
            if (row == _rowCount || column == _columnCount)
                return;

            IBaseItem ibi = _cells[column + row * _columnCount].GetItem();
            if (ibi != null)
            {
                base.RemoveItem(ibi);
                _cells[column + row * _columnCount].SetItem(null);
            }
        }

        private Object Locker = new Object();
        public override void Clear()
        {
            Monitor.Enter(Locker);
            try
            {
                List<IBaseItem> list = GetItems();
                while (list.Count > 0)
                    RemoveItem(list.First());
            }
            catch (Exception ex)
            {
                Console.WriteLine("Method - Clear");
                Console.WriteLine(ex.StackTrace);
            }
            finally
            {
                Monitor.Exit(Locker);
            }
        }

        //overrides
        protected internal override bool GetHoverVerification(float xpos, float ypos)
        {
            return false;
        }

        /// <summary>
        /// Add item to the Grid
        /// </summary>
        public override void AddItem(IBaseItem item)
        {
            //ignore if it is out of space, add in free cell, attach row and collumn numbers
            foreach (Cell cell in _cells)
            {
                if (cell.GetItem() == null)
                {
                    base.AddItem(item);
                    cell.SetItem(item);
                    UpdateLayout();
                    return;
                }
            }
        }

        /// <summary>
        /// Insert item to the Grid by row and column number
        /// </summary>
        public void InsertItem(IBaseItem item, int row, int column)
        {
            if (row == _rowCount || column == _columnCount)
                return;
            base.AddItem(item);
            //_cells[row + column * _rowCount].SetItem(item);

            RemoveItem(row, column);

            _cells[column + row * _columnCount].SetItem(item);
            UpdateLayout();
        }

        public override void InsertItem(IBaseItem item, int index)
        {
            if (_columnCount == 0)
                return;
            int row, column;
            row = index / _columnCount;
            column = index - row * _columnCount;
            InsertItem(item, row, column);
        }

        /// <summary>
        /// Set width of the Grid
        /// </summary>
        public override void SetWidth(int width)
        {
            base.SetWidth(width);
            UpdateLayout();
        }

        /// <summary>
        /// Set height of the Grid
        /// </summary>
        public override void SetHeight(int height)
        {
            base.SetHeight(height);
            UpdateLayout();
        }

        /// <summary>
        /// Set X position of the Grid
        /// </summary>
        public override void SetX(int _x)
        {
            base.SetX(_x);
            UpdateLayout();
        }

        /// <summary>
        /// Set Y position of the Grid
        /// </summary>
        public override void SetY(int _y)
        {
            base.SetY(_y);
            UpdateLayout();
        }

        //TMP
        private Int32[] colWidth;
        private Int32[] rowHeight;
        internal Int32[] GetColWidth()
        {
            return colWidth;
        }
        internal Int32[] GetRowHeight()
        {
            return rowHeight;
        }

        private bool _isUpdating = false;
        /// <summary>
        /// Update all children and grid sizes and positions
        /// according to confines
        /// </summary>
        //Update Layout
        public void UpdateLayout()
        {
            // Console.WriteLine("grid update");
            /*
             * Алгоритм:
             * 1. найти максимумы по ширине в столбцах
             * 2. запомнить размеры по ширине каждой ячейки
             * 3. найти все максимумы по высоте в строках
             * 4. запомнить размеры по высоте каждой ячейки
             * 5. цикл по каждой ячейке
             * 6. установть размеры ячейки
             * 7. разместить элемент в ячейке
             * 8. повторить цикл с пункта 5, пока есть ячейки
             * 9. готово, пить кофе
             */

            /*
             * Учесть
             * У клетки минимальную/максимальную длину/ширину
             * Выравнивание элемента
             * Политику растягивания размеров
             */
            List<IBaseItem> list = GetItems();
            if (list == null || list.Count == 0 || _isUpdating)
                return;
            _isUpdating = true;

            //1, 2
            Int32[] columns_width = GetColumnsWidth();
            colWidth = columns_width;

            //3, 4
            Int32[] rows_height = GetRowsHeight();
            rowHeight = rows_height;

            //5
            int x_offset = 0;
            int y_offset = 0;
            for (int r = 0; r < _rowCount; r++)
            {
                int index = 0;
                for (int c = 0; c < _columnCount; c++)
                {
                    // index = r + c * _rowCount;
                    index = c + r * _columnCount;

                    IBaseItem item = _cells[index].GetItem();
                    /*if (item == null)
                        continue;*/

                    _cells[index].SetRow(r);
                    _cells[index].SetColumn(c);

                    //6
                    _cells[index].SetWidth(columns_width[c]);
                    _cells[index].SetHeight(rows_height[r]);
                    _cells[index].SetX(GetX() + GetPadding().Left + x_offset);
                    _cells[index].SetY(GetY() + GetPadding().Top + y_offset);

                    if (item != null)
                    {
                        if (item.GetWidthPolicy().HasFlag(SizePolicy.Expand))
                            item.SetWidth(columns_width[c] - item.GetMargin().Left - item.GetMargin().Right);

                        if (item.GetHeightPolicy().HasFlag(SizePolicy.Expand))
                            item.SetHeight(rows_height[r] - item.GetMargin().Top - item.GetMargin().Bottom);
                        item.SetConfines();
                    }

                    //7
                    _cells[index].UpdateBehavior();

                    x_offset += _cells[index].GetWidth() + GetSpacing().Horizontal;
                }
                y_offset += _cells[index].GetHeight() + GetSpacing().Vertical;
                x_offset = 0;
            }
            _isUpdating = false;
        }

        private Int32[] GetRowsHeight()
        {
            Int32[] rows_height = new Int32[_rowCount];
            List<int[]> list_height = new List<int[]>();

            int total_space = GetHeight() - GetPadding().Top - GetPadding().Bottom;
            int free_space = total_space;
            int prefer_height = (total_space - GetSpacing().Vertical * (_rowCount - 1)) / _rowCount;
            int count = _rowCount;

            for (int r = 0; r < _rowCount; r++)
            {
                for (int c = 0; c < _columnCount; c++)
                {
                    IBaseItem item = _cells[c + r * _columnCount].GetItem();

                    if (item == null || !item.IsVisible() || !item.IsDrawable())
                    {
                        list_height.Add(new int[2] { r, -1 });
                        continue;
                    }

                    if (item.GetHeightPolicy() == SizePolicy.Fixed)
                    {
                        list_height.Add(new int[2] { r, item.GetHeight() + item.GetMargin().Top + item.GetMargin().Bottom });
                    }
                    else
                    {
                        list_height.Add(new int[2] { r, 0 });
                    }
                }
            }
            ///////////
            List<int[]> m_height = new List<int[]>();
            for (int r = 0; r < _rowCount; r++)
            {
                int max = -10;
                for (int c = 0; c < _columnCount; c++)
                {
                    if (list_height[c + r * _columnCount][1] > max)
                        max = list_height[c + r * _columnCount][1];
                }
                m_height.Add(new int[2] { r, max });
                if (max == -1)
                {
                    count--;
                    if (count == 0)
                        count++;
                    prefer_height = (free_space - GetSpacing().Vertical * (count - 1)) / count;
                }
            }

            m_height.Sort((x, y) => y[1].CompareTo(x[1]));

            foreach (var pair in m_height)
            {
                if (pair[1] == 0)
                    pair[1] = prefer_height;
                else if (pair[1] < 0)
                {
                    pair[1] = 0;
                }
                else
                {
                    free_space -= pair[1];
                    count--;
                    if (count == 0)
                        count++;
                    prefer_height = (free_space - GetSpacing().Vertical * (count - 1)) / count;
                }
            }

            m_height.Sort((x, y) => x[0].CompareTo(y[0]));

            for (int i = 0; i < rows_height.Length; i++)
                rows_height[i] = m_height[i][1];

            return rows_height;
        }

        private Int32[] GetColumnsWidth()
        {
            Int32[] columns_width = new Int32[_columnCount];
            List<int[]> list_width = new List<int[]>();

            int total_space = GetWidth() - GetPadding().Left - GetPadding().Right;
            int free_space = total_space;
            //int prefer_width = (int)Math.Round((float)(total_space - GetSpacing().Horizontal * (_columnCount - 1)) / (float)_columnCount, MidpointRounding.AwayFromZero);
            int prefer_width = (total_space - GetSpacing().Horizontal * (_columnCount - 1)) / _columnCount;
            int count = _columnCount;

            for (int c = 0; c < _columnCount; c++)
            {
                for (int r = 0; r < _rowCount; r++)
                {
                    IBaseItem item = _cells[c + r * _columnCount].GetItem();
                    if (item == null || !item.IsVisible() || !item.IsDrawable())
                    {
                        list_width.Add(new int[2] { c, -1 });
                        continue;
                    }

                    if (item.GetWidthPolicy() == SizePolicy.Fixed)
                    {
                        list_width.Add(new int[2] { c, item.GetWidth() + item.GetMargin().Left + item.GetMargin().Right });
                    }
                    else
                    {
                        list_width.Add(new int[2] { c, 0 });
                    }
                }
            }
            //////////
            List<int[]> m_width = new List<int[]>();
            for (int c = 0; c < _columnCount; c++)
            {
                int max = -10;
                for (int r = 0; r < _rowCount; r++)
                {
                    if (list_width[r + c * _rowCount][1] > max)
                        max = list_width[r + c * _rowCount][1];
                }
                m_width.Add(new int[2] { c, max });
                if (max == -1)
                {
                    count--;
                    if (count == 0)
                        count++;
                    prefer_width = (free_space - GetSpacing().Horizontal * (count - 1)) / count;
                }
            }

            m_width.Sort((x, y) => y[1].CompareTo(x[1]));

            foreach (var pair in m_width)
            {
                if (pair[1] == 0)
                    pair[1] = (int)prefer_width;
                else if (pair[1] < 0)
                {
                    pair[1] = 0;
                }
                else
                {
                    free_space -= pair[1];
                    count--;
                    if (count == 0)
                        count++;
                    prefer_width = (free_space - GetSpacing().Horizontal * (count - 1)) / count;
                }
            }

            m_width.Sort((x, y) => x[0].CompareTo(y[0]));

            for (int i = 0; i < columns_width.Length; i++)
                columns_width[i] = m_width[i][1];

            return columns_width;
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/Grid.java
package com.spvessel;

import com.spvessel.Common.DefaultsService;
import com.spvessel.Core.Cell;
import com.spvessel.Core.InterfaceBaseItem;
import com.spvessel.Core.InterfaceGrid;
import com.spvessel.Flags.SizePolicy;

import java.util.List;
import java.util.LinkedList;

public class Grid extends Prototype implements InterfaceGrid {
    private static int count = 0;

    /**
     * Constructs a Grid
     */
    public Grid() {
        setItemName("Grid_" + count);
        count++;

        // setStyle(DefaultsService.getDefaultStyle("SpaceVIL.Grid"));
        setStyle(DefaultsService.getDefaultStyle(Grid.class));
        isFocusable = false;
    }

    /**
     * Constructs a Grid with the given rows and columns
     */
    public Grid(int rows, int columns) {
        this();
        _row_count = rows;
        _column_count = columns;
        initCells();
    }

    // rows and counts
    private List<Cell> _cells;

    void initCells() {
        _cells = new LinkedList<>();
        for (int i = 0; i < _row_count; i++) {
            for (int j = 0; j < _column_count; j++) {
                _cells.add(new Cell(this, i, j));
            }
        }
    }

    private int _row_count = 1;

    /**
     * Set count of the rows
     */
    public void setRowCount(int capacity) {
        if (capacity != _row_count)
            _row_count = capacity;
    }

    public int getRowCount() {
        return _row_count;
    }

    private int _column_count = 1;

    /**
     * Set count of the columns
     */
    public void setColumnCount(int capacity) {
        if (capacity != _column_count)
            _column_count = capacity;
        // Need to initCells REFACTOR!
    }

    public int getColumnCount() {
        return _column_count;
    }

    /**
     * Returns the cell by row and column number
     */
    public Cell getCell(int row, int column) {
        Cell cell = null;
        try {
            cell = _cells.get(column + row * _column_count);
        } catch (Exception ex) {
            System.out.println("Cells row and column out of range.\n" + ex.toString());
            return cell;
        }
        return cell;
    }

    /**
     * Returns all cells list
     */
    public List<Cell> getAllCells() {
        return _cells;
    }

    // overrides
    @Override
    public boolean getHoverVerification(float xpos, float ypos) {
        return false;
    }

    /**
     * Add item to the Grid
     */
    @Override
    public void addItem(InterfaceBaseItem item) {
        // ignore if it is out of space, add in free cell, attach row and collumn
        // numbers
        for (Cell cell : _cells) {
            if (cell.getItem() == null) {
                super.addItem(item);
                cell.setItem(item);
                updateLayout();
                return;
            }
        }
    }

    /**
     * Insert item to the Grid by row and column number
     */
    public void insertItem(InterfaceBaseItem item, int row, int column) {
        super.addItem(item);
        // _cells[row + column * _row_count].setItem(item);
        // Console.WriteLine(column + row * _column_count);
        _cells.get(column + row * _column_count).setItem(item);
        updateLayout();
    }

    /**
     * Set width of the Grid
     */
    @Override
    public void setWidth(int width) {
        super.setWidth(width);
        updateLayout();
    }

    /**
     * Set X position of the Grid
     */
    @Override
    public void setX(int _x) {
        super.setX(_x);
        updateLayout();
    }

    /**
     * Set height of the Grid
     */
    @Override
    public void setHeight(int height) {
        super.setHeight(height);
        updateLayout();
    }

    /**
     * Set Y position of the Grid
     */
    @Override
    public void setY(int _y) {
        super.setY(_y);
        updateLayout();
    }

    // TMP
    private int[] colWidth;
    private int[] rowHeight;

    int[] getColWidth() {
        return colWidth;
    }

    int[] getRowHeight() {
        return rowHeight;
    }

    /**
     * Update all children and grid sizes and positions
     * according to confines
     */
    // update Layout
    public void updateLayout() {

        if (getItems().size() == 0)
            return;

        int[] columns_width = getColumnsWidth();
        colWidth = columns_width;

        int[] rows_height = getRowsHeight();
        rowHeight = rows_height;

        int x_offset = 0;
        int y_offset = 0;
        for (int r = 0; r < _row_count; r++) {
            int index = 0;
            for (int c = 0; c < _column_count; c++) {
                index = c + r * _column_count;

                InterfaceBaseItem item = _cells.get(index).getItem();

                _cells.get(index).setRow(r);
                _cells.get(index).setColumn(c);

                // 6
                _cells.get(index).setWidth(columns_width[c]);
                _cells.get(index).setHeight(rows_height[r]);
                _cells.get(index).setX(getX() + getPadding().left + x_offset);
                _cells.get(index).setY(getY() + getPadding().top + y_offset);

                if (item != null) {
                    if (item.getWidthPolicy() == SizePolicy.EXPAND)
                        item.setWidth(columns_width[c] - item.getMargin().left - item.getMargin().right);

                    if (item.getHeightPolicy() == SizePolicy.EXPAND)
                        item.setHeight(rows_height[r] - item.getMargin().top - item.getMargin().bottom);
                    item.setConfines();
                }

                // 7
                _cells.get(index).updateBehavior();
                x_offset += _cells.get(index).getWidth() + getSpacing().horizontal;
            }
            y_offset += _cells.get(index).getHeight() + getSpacing().vertical;
            x_offset = 0;
        }

    }

    private int[] getRowsHeight() {
        int[] rows_height = new int[_row_count];
        List<int[]> list_height = new LinkedList<>();

        int total_space = getHeight() - getPadding().top - getPadding().bottom;
        int free_space = total_space;
        int prefer_height = (total_space - getSpacing().vertical * (_row_count - 1)) / _row_count;
        int count = _row_count;

        for (int r = 0; r < _row_count; r++) {
            for (int c = 0; c < _column_count; c++) {
                InterfaceBaseItem item = _cells.get(c + r * _column_count).getItem();

                if (item == null || !item.isVisible() || !item.isDrawable()) {
                    list_height.add(new int[] { r, -1 });
                    continue;
                }

                if (item.getHeightPolicy() == SizePolicy.FIXED) {
                    list_height.add(new int[] { r, item.getHeight() + item.getMargin().top + item.getMargin().bottom });
                } else {
                    list_height.add(new int[] { r, 0 });
                }
            }
        }
        ///////////
        List<Integer[]> m_height = new LinkedList<>();
        for (int r = 0; r < _row_count; r++) {
            int max = -10;
            for (int c = 0; c < _column_count; c++) {
                if (list_height.get(c + r * _column_count)[1] > max)
                    max = list_height.get(c + r * _column_count)[1];
            }
            m_height.add(new Integer[] { r, max });
            if (max == -1) {
                count--;
                if (count == 0)
                    count++;
                prefer_height = (free_space - getSpacing().vertical * (count - 1)) / count;
            }
        }

        m_height.sort((li1, li2) -> li2[1].compareTo(li1[1]));

        for (Integer[] pair : m_height) {
            if (pair[1] == 0)
                pair[1] = prefer_height;
            else if (pair[1] < 0) {
                pair[1] = 0;
            } else {
                free_space -= pair[1];
                count--;
                if (count == 0)
                    count++;
                prefer_height = (free_space - getSpacing().vertical * (count - 1)) / count;
            }
        }

        m_height.sort((li1, li2) -> li1[0].compareTo(li2[0]));

        for (int i = 0; i < _row_count; i++)
            rows_height[i] = m_height.get(i)[1];

        return rows_height;
    }

    private int[] getColumnsWidth() {
        int[] columns_width = new int[_column_count];
        List<int[]> list_width = new LinkedList<>();

        int total_space = getWidth() - getPadding().left - getPadding().right;
        int free_space = total_space;
        int prefer_width = (total_space - getSpacing().horizontal * (_column_count - 1)) / _column_count;
        int count = _column_count;

        for (int c = 0; c < _column_count; c++) {
            for (int r = 0; r < _row_count; r++) {
                InterfaceBaseItem item = _cells.get(c + r * _column_count).getItem();
                if (item == null || !item.isVisible() || !item.isDrawable()) {
                    list_width.add(new int[] { c, -1 });
                    continue;
                }

                if (item.getWidthPolicy() == SizePolicy.FIXED) {
                    list_width.add(new int[] { c, item.getWidth() + item.getMargin().left + item.getMargin().right });
                } else {
                    list_width.add(new int[] { c, 0 });
                }
            }
        }
        //////////
        List<Integer[]> m_width = new LinkedList<>();
        for (int c = 0; c < _column_count; c++) {
            int max = -10;
            for (int r = 0; r < _row_count; r++) {
                if (list_width.get(r + c * _row_count)[1] > max)
                    max = list_width.get(r + c * _row_count)[1];
            }
            m_width.add(new Integer[] { c, max });
            if (max == -1) {
                count--;
                if (count == 0)
                    count++;
                prefer_width = (free_space - getSpacing().horizontal * (count - 1)) / count;
            }
        }

        m_width.sort((li1, li2) -> li2[1].compareTo(li1[1]));

        for (Integer[] pair : m_width) {
            if (pair[1] == 0)
                pair[1] = prefer_width;
            else if (pair[1] < 0) {
                pair[1] = 0;
            } else {
                free_space -= pair[1];
                count--;
                if (count == 0)
                    count++;
                prefer_width = (free_space - getSpacing().horizontal * (count - 1)) / count;
            }
        }

        m_width.sort((li1, li2) -> li1[0].compareTo(li2[0]));

        for (int i = 0; i < _column_count; i++)
            columns_width[i] = m_width.get(i)[1];

        return columns_width;
    }
}